ngx_http_sxg_filter_module.c

```diff
@@ -303,38 +333,123 @@ static ngx_str_t extract_angled_url(char* str, size_t len) {
     ngx_str_t url = extract_angled_url(str, tail);
     char* param = str;
+    bool found = false;
+    ngx_subresource_t* new_preload = NULL;
+    ngx_str_t as = ngx_null_string;
+
     while (param < str + tail) {
       const size_t rest = str + tail - param;
       size_t param_tail = get_term_length(param, rest, ';', "<>");
+
       if (param_is_preload(param, param_tail)) {
-        ngx_str_t* const new_url = ngx_array_push(dst);
-        *new_url = url;
-        break;
+        if (dst != NULL) {
+          new_preload = ngx_array_push(dst);
+          new_preload->url.data = ngx_palloc(r->pool, url.len);
+          new_preload->url.len = url.len;
+          ngx_memcpy(new_preload->url.data, url.data, url.len);
+          while (new_preload->url.data[new_preload->url.len - 1] == ' ') {
+            new_preload->url.len--;
+          }
+          if (as.data != NULL) {
+            new_preload->as = as;
+          }
+        }
+        found = true;
+      }
+      if (param_is_as(param, param_tail, (const char**)&as.data, &as.len)) {
+        if (found && new_preload != NULL) {
+          new_preload->as = as;
+          ngx_log_error(
+              NGX_LOG_DEBUG, r->connection->log, 0,
+              "nginx-sxg-module: found as statement in link preload header: %V",
+              &new_preload->as);
+        }
       }
       param += param_tail + 1;
     }
+    if (!found) {
+      if (non_preload_headers == NULL) {
+        if (non_preload_headers_len > 0) {
+          ++non_preload_headers_len;
+        }
+        non_preload_headers_len += tail;
+      } else {
+        if (non_preload_headers->len > 0) {
+          ngx_memcpy(non_preload_headers->data + non_preload_headers_len, ",",
+                     1);
+          ++non_preload_headers_len;
+        }
+        ngx_memcpy(non_preload_headers->data + non_preload_headers_len, str,
+                   tail);
+        non_preload_headers_len += tail;
+      }
+    }
     str += tail + 1;
   }
+
+  return non_preload_headers_len;
 }
 
-static ngx_array_t* get_preload_list(ngx_str_t* link, ngx_http_request_t* req) {
-  ngx_array_t* const urls = ngx_array_create(req->pool, 1, sizeof(ngx_str_t));
-  extract_preaload_url_list(link, urls, req);
+static ngx_array_t* get_preload_list(ngx_str_t* link,
+                                     ngx_str_t* non_preload_headers,
+                                     ngx_http_request_t* req) {
+  ngx_array_t* const urls =
+      ngx_array_create(req->pool, 1, sizeof(ngx_subresource_t));
+  size_t non_preload_headers_len =
+      extract_preload_url_list(link, NULL, NULL, req);
+  non_preload_headers->data = ngx_palloc(req->pool, non_preload_headers_len);
+  non_preload_headers->len = non_preload_headers_len;
+  extract_preload_url_list(link, urls, non_preload_headers, req);
   return urls;
 }
 
+static bool set_str(ngx_pool_t* pool, ngx_str_t* dst, const char* src) {
+  const size_t len = strlen(src);
+  dst->data = ngx_palloc(pool, len);
```

Comment:
Do we need to [`ngx_pfree()`](https://www.nginx.com/resources/wiki/extending/api/alloc/#ngx-pfree) the old value if non-NULL?